### PR TITLE
Implementation of PassLocalVarType

### LINKED ISSUE #669: Assign location type to Local Var

### PR DESCRIPTION
**Pull request by Stagno**
## Technical Description

- Add `LocalVariableData` (data container for a local variable) and `LocalVariableType` which can be `Scalar, OnCells, OnEdges, OnVertices, OnIJ`.
- Add pass (and unit tests) `PassLocalVariableType` to determine the type of a local variable.
- Add map from access ID of variable to `LocalVariableData` in metadata (not serialized).
- Add helper method `StencilMetaInformation::declareVar()` to declare a variable (same as `StencilMetaInformation::addStmt()` but doesn't require a `VarDeclStmt` in input).
- In `GridTypeChecker` check if all variable types match the grid type.

### Resolves / Enhances

resolves #669

### CODE REVIEW

**Comment by Stagno** [comment]
launch jenkins

**Comment by Stagno** [comment]
launch jenkins

**Comment by Stagno** [comment]
launch jenkins

File: dawn/src/dawn/Optimizer/PassInlining.cpp [hunk 1]
```diff
@@ -117,23 +117,20 @@ class Inliner : public iir::ASTVisitor {
...
       // return value in a local variable.
-      int AccessID = instantiation_->nextUID();
-      auto returnVarName = iir::InstantiationHelper::makeLocalVariablename(
-          curStencilFunctioninstantiation_->getName(), AccessID);
 
-      newExpr_ = std::make_shared<iir::VarAccessExpr>(returnVarName);
-      auto newStmt =
-          iir::makeVarDeclStmt(dawn::Type(BuiltinTypeID::Float, CVQualifier::Const), returnVarName,
-                               0, "=", std::vector<std::shared_ptr<iir::Expr>>{stmt->getExpr()});
+      // Declare and register the variable
```
Comment (mroethlin) [inline dawn/src/dawn/Optimizer/PassInlining.cpp]: I'd prefer passing a local var `keepVarName = false` instead of just `false` here, but maybe that's overly verbose.

File: dawn/src/dawn/Optimizer/PassLocalVarType.cpp [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+#include "dawn/IIR/ASTExpr.h"
+#include "dawn/IIR/ASTStmt.h"
+#include "dawn/IIR/DoMethod.h"
+#include "dawn/IIR/StencilInstantiation.h"
+#include "dawn/IIR/StencilMetaInformation.h"
+
+namespace dawn {
+namespace {
+
+// Resets types of all variables to "not computed" (type_ = std::nullopt)
```
Comment (mroethlin) [inline dawn/src/dawn/Optimizer/PassLocalVarType.cpp]: I think I might prefer the metadata having a `resetVarTypes` function instead of allowing the whole world to arbitrarily change individual `LocalVariableData` information

File: dawn/src/dawn/Optimizer/PassLocalVarType.cpp [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+      visitedSet.insert(dependee);
+      // If no other variable depends on the dependee, return
+      if(!dependencyMap_.count(dependee)) {
+        return;
+      }
+      // Collect the type to propagate
+      const iir::LocalVariableData& dependeeData =
+          metadata_.getLocalVariableDataFromAccessID(dependee);
+      // If the dependee's type is not set yet, it has been accessed before being declared.
+      DAWN_ASSERT_MSG(dependeeData.isTypeSet(), "Variable accessed before being declared.");
```
Comment (mroethlin) [inline dawn/src/dawn/Optimizer/PassLocalVarType.cpp]: is there a good reason that we assert here, but throw on 58?

File: dawn/src/dawn/Optimizer/PassLocalVarType.cpp [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+        throw std::runtime_error(dawn::format(
+            "Variable assignment inside rhs of variable assignment is not supported. Line %d.",
+            expr->getSourceLocation().Line));
+      }
+      // Assume scalar if nothing proves the opposite during the visit
+      curVarID_ = iir::getAccessID(expr->getLeft());
+      updateVariableType(*curVarID_, iir::LocalVariableType::Scalar, expr->getSourceLocation());
+      // If the set of dependers for this variable is not there, it means the variable hasn't been
+      // declared
+      DAWN_ASSERT_MSG(dependencyMap_.count(*curVarID_), "Variable accessed before being declared.");
```
Comment (mroethlin) [inline dawn/src/dawn/Optimizer/PassLocalVarType.cpp]: c.f. above

File: dawn/src/dawn/Optimizer/TemporaryHandling.cpp [hunk 5]
```diff
@@ -146,10 +146,12 @@ void demoteTemporaryFieldToLocalVariable(iir::StencilInstantiation* instantiatio
...
-  // Create the new `VarDeclStmt` which will replace the old `ExprStmt`
-  std::shared_ptr<iir::Stmt> varDeclStmt =
-      iir::makeVarDeclStmt(Type(BuiltinTypeID::Float), varname, 0, "=",
-                           std::vector<std::shared_ptr<iir::Expr>>{assignmentExpr->getRight()});
+  // Remove the field
+  instantiation->getMetaData().removeAccessID(AccessID);
+
+  // Create the new `VarDeclStmt` which will replace the old `ExprStmt` and register the variable
+  std::shared_ptr<iir::Stmt> varDeclStmt = instantiation->getMetaData().declareVar(
+      true, varname, Type(BuiltinTypeID::Float), assignmentExpr->getRight(), AccessID);
```
Comment (mroethlin) [inline dawn/src/dawn/Optimizer/TemporaryHandling.cpp]: so I'm not sure I understand the keep var name thing. Why is it true here, but false in other instances you `declareVar`? is this related to the (broken) `keepVarNames` flag?

File: dawn/src/dawn/Optimizer/PassInlining.cpp (hunk 1 above)
Comment (Stagno) [reply dawn/src/dawn/Optimizer/PassInlining.cpp]: Agree, makes the code more readable.

File: dawn/src/dawn/Optimizer/PassLocalVarType.cpp (hunk 2 above)
Comment (Stagno) [reply dawn/src/dawn/Optimizer/PassLocalVarType.cpp]: Ok.

**Comment by Stagno** [comment]
launch jenkins
